utils/background-task-queue.ts

```diff
@@ -0,0 +1,232 @@
+import path from "path";
+import EventEmitter from "events";
+import { Worker, isMainThread, parentPort } from "worker_threads";
+import { mkdir, writeFile } from "fs/promises";
+
+import { nanoid } from "nanoid";
+import split2 from "split2";
+
+import { legacyFilename, env } from "./misc.js";
+
+type Message = { id: string };
+
+interface OperationRequest extends Message {
+  type: "init";
+  modulePath: string;
+}
+interface CallRequest extends Message {
+  type: "call";
+  input?: unknown;
+}
+type Request = OperationRequest | CallRequest;
+
+interface Response extends Message {
+  type: "success" | "failure";
+  result: unknown;
+}
+
+if (!isMainThread) {
+  let task: (input: CallRequest["input"]) => Promise<unknown>;
+  parentPort.addListener("message", async (req: Request) => {
+    const { id } = req;
+
+    if (req.type === "init") {
+      const mod = await import(req.modulePath);
+      task = mod.default;
+      return parentPort.postMessage({ id, result: "success" });
+    }
+
+    if (req.type === "call") {
+      try {
+        const result = await task(req.input);
+        const msg: Response = { id, type: "success", result };
+        return parentPort.postMessage(msg);
+      } catch (error) {
+        const msg: Response = { id, type: "failure", result: error };
+        return parentPort.postMessage(msg);
+      }
+    }
+  });
+}
+
+class Lock {
+  private isLocked = false;
+  private emitter = new EventEmitter();
+
+  async acquire() {
+    while (this.isLocked) {
```

Comment:
Can this loop run more than once? It looks like it "unlock"s only on release()... if so: 
  
```suggestion
    if (this.isLocked) {
```

Reply:
Ah. Now I remember. If we use `if`, multiple requests can start at once when the lock is released (the promise resolves at once). The `while` loop disallows it (I don't fully understand how, but that's how it works).

Reply:
updated suggestion to include what you said.